### ISSUE TITLE
clang with -Wnull-dereference doesn't produce expected warning

### ISSUE THREAD

**Issue by tklengyel** [llvm/llvm-project#113068]
Greetings,
while compiling the following program with `-Wnull-dereference` no warning is produced by clang (`Ubuntu clang version 20.0.0 (++20241009053718+c198f775cdcd-1~exp1~20241009053853.1977`):
```c
#include <stdio.h>

int main(int argc, char **argv)
{
    int *yikes = NULL;
    *yikes = 1;

    fprintf(stdout, "Hello world!\n");

    return *yikes;
}
```

Expected result would be like what you get with gcc (although it only seems to work with gcc if you have optimizations enabled and not with -O0):
```
$ gcc -O1 -Wnull-dereference -o yikes yikes.c
yikes.c: In function ‘main’:
yikes.c:8:12: warning: null pointer dereference [-Wnull-dereference]
    8 |     *yikes = 1;
      |     ~~~~~~~^~~
yikes.c:12:12: warning: null pointer dereference [-Wnull-dereference]
   12 |     return *yikes;
      |            ^~~~~~
```

**Comment by shafik** [comment]
Godbolt: https://godbolt.org/z/4bv5a35TE

**Comment by a-tarasyuk** [comment]
what diagnostic should be used in this case? should it be one of these, or is there a need for a new one?

https://github.com/llvm/llvm-project/blob/34d4f660fe57132d17d2e37b72ccfc1d07269de9/clang/lib/Sema/SemaExpr.cpp#L578

https://github.com/llvm/llvm-project/blob/34d4f660fe57132d17d2e37b72ccfc1d07269de9/clang/lib/Sema/SemaInit.cpp#L7522

**Comment by shafik** [comment]
I think we could possibly make `warn_indirection_through_null` via `CheckForNullPointerDereference` but I don't think the diagnostic would be accurate b/c I am guessing the optimizer won't have the same information or won't want to do the extra work.

CC @AaronBallman

**Comment by AaronBallman** [comment]
The issue is that `CheckForNullPointerDereference()` is only run during initialization, and this is not initialization but assignment. In order to support this, we need to build a CFG, and that's an expensive operation that we would typically only do as an opt-in analysis diagnostic or through the static analyzer. And the static analyzer does catch this: https://godbolt.org/z/96K169hEe

**Comment by a-tarasyuk** [comment]
@AaronBallman thank you for the explanation.

**Comment by eyalroz** [comment]
The minimal example can be [simplified further](https://godbolt.org/z/95neocf71)...:
```
int main() {
    int* ptr = nullptr;
    return *ptr + 1;
}
```
This has just come up in [this StackOverflow question](https://stackoverflow.com/q/79662603/1593077).
